src/macro-expansion.md

```diff
@@ -3,147 +3,169 @@
-  the macro is ambiguous.
-
-The full interface is defined [here][code_parse_int].
-
-The macro parser does pretty much exactly the same as a normal regex parser with
-one exception: in order to parse different types of metavariables, such as
-`ident`, `block`, `expr`, etc., the macro parser must sometimes call back to the
-normal Rust parser.
-
-As mentioned above, both definitions and invocations of macros are parsed using
-the macro parser. This is extremely non-intuitive and self-referential. The code
-to parse macro _definitions_ is in
-[`src/librustc_expand/mbe/macro_rules.rs`][code_mr]. It defines the pattern for
-matching for a macro definition as `$( $lhs:tt => $rhs:tt );+`. In other words,
-a `macro_rules` definition should have in its body at least one occurrence of a
-token tree followed by `=>` followed by another token tree. When the compiler
-comes to a `macro_rules` definition, it uses this pattern to match the two token
-trees per rule in the definition of the macro _using the macro parser itself_.
-In our example definition, the metavariable `$lhs` would match the patterns of
-both arms: `(print $mvar:ident)` and `(print twice $mvar:ident)`.  And `$rhs`
-would match the bodies of both arms: `{ println!("{}", $mvar); }` and `{
-println!("{}", $mvar); println!("{}", $mvar); }`. The parser would keep this
-knowledge around for when it needs to expand a macro invocation.
-
-When the compiler comes to a macro invocation, it parses that invocation using
-the same NFA-based macro parser that is described above. However, the matcher
-used is the first token tree (`$lhs`) extracted from the arms of the macro
-_definition_. Using our example, we would try to match the token stream `print
-foo` from the invocation against the matchers `print $mvar:ident` and `print
-twice $mvar:ident` that we previously extracted from the definition.  The
-algorithm is exactly the same, but when the macro parser comes to a place in the
-current matcher where it needs to match a _non-terminal_ (e.g. `$mvar:ident`),
-it calls back to the normal Rust parser to get the contents of that
-non-terminal. In this case, the Rust parser would look for an `ident` token,
-which it finds (`foo`) and returns to the macro parser. Then, the macro parser
-proceeds in parsing as normal. Also, note that exactly one of the matchers from
-the various arms should match the invocation; if there is more than one match,
-the parse is ambiguous, while if there are no matches at all, there is a syntax
-error.
-
-For more information about the macro parser's implementation, see the comments
-in [`src/librustc_expand/mbe/macro_parser.rs`][code_mp].
+foo!(bar!(baz));
+```
 
-### Hygiene
+A lazy expansion would expand `foo!` first. An eager expansion would expand
+`bar!` first. Implementing eager expansion more generally would be challenging,
+but we implement it for a few special built-in macros for the sake of user
```

Comment:
My understanding is that because they only take literals, expanding early produces better UX (but tbh, I don't know how off the top of my head).

I tried to reword to make that a bit clearer.